src/fuse_fs.c

```diff
@@ -0,0 +1,240 @@
+	if (strcmp(path, "/") == 0) {
+		stbuf->st_mode = S_IFDIR | 0755;
+		stbuf->st_nlink = 2;
+	}
+	else {
+		char fpath[PATH_MAX];
+		get_file_path(fpath, fname);
+		fprintf(stderr, "FPath: %s\n", fpath);
+		res = lstat(fpath, stbuf);
+	}
+
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int altfs_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
+				off_t offset, struct fuse_file_info *fi,
+				enum fuse_readdir_flags flags)
+{
+	fprintf(stderr, "Inside readdir\n");
+	fprintf(stderr, "Path: %s\n", path);
+	int retstat = 0;
+	DIR *dp;
+	struct dirent *de;
+
+	filler(buf, ".", NULL, 0, 0);
+	filler(buf, "..", NULL, 0, 0);
+
+	dp = (DIR *) (uintptr_t) fi->fh;
+
+	if (dp != NULL) {
+		fprintf(stderr, "attempting to read dir\n");
+		de = readdir(dp);
+		fprintf(stderr, "got first readdir\n");
+		if (de == 0)
+		{
+			fprintf(stderr, "error in reading dir");
+			return -errno;
+		}
+
+		do {
+			fprintf(stderr, "adding entry %s", de->d_name);
+			if (filler(buf, de->d_name, NULL, 0, 0) != 0) {
+				fprintf(stderr, "buffer full");
+				return -ENOMEM;
+			}
+		} while ((de = readdir(dp)) != NULL);
+	} else {
+		fprintf(stderr, "dp is null\n");
+	}
+
+	return retstat;
+}
+
+static int altfs_create(const char *path, mode_t mode, struct fuse_file_info *fi)
+{
+	fprintf(stderr, "Inside create\n");
+	fprintf(stderr, "Path: %s\n", path);
+	int res;
+
+	char fpath[PATH_MAX];
+	get_file_path(fpath, fname);
+	fprintf(stderr, "FPath: %s\n", fpath);
+
+	res = creat(fpath, mode);
+	fprintf(stderr, "Result of create: %d\n", res);
+	if (res == -1)
+		return -errno;
+
+	fi->fh = res;
+	return 0;
+}
+
+static int altfs_open(const char *path, struct fuse_file_info *fi)
+{
+	fprintf(stderr, "Inside open\n");
+	fprintf(stderr, "Path: %s\n", path);
+	int res;
+
+	char fpath[PATH_MAX];
+	get_file_path(fpath, fname);
+	fprintf(stderr, "FPath: %s\n", fpath);
+
+	fprintf(stderr, "attempting to open\n");
+	res = open(fpath, fi->flags);
+	fprintf(stderr, "Result: %d\n", res);
+	if (res < 0) {
+		fprintf(stderr, "error in opening!\n");
+		return -errno;
+	}
+
+	fi->fh = res;
+
+	return 0;
+}
+
+static int altfs_truncate(const char *path, off_t size, struct fuse_file_info *fi)
+{
+	fprintf(stderr, "Inside truncate\n");
+	fprintf(stderr, "Path: %s\n", path);
+	int res;
+
+	char fpath[PATH_MAX];
+	get_file_path(fpath, fname);
+	fprintf(stderr, "FPath: %s\n", fpath);
+
+	if (fi != NULL)
+		res = ftruncate(fi->fh, size);
+	else
+		res = truncate(fpath, size);
+
+	if (res == -1)
+		return -errno;
+
+	return 0;
+}
+
+static int altfs_write(const char* path, const char *buf, size_t size, off_t offset, struct fuse_file_info* fi)
+{
+	fprintf(stderr, "Inside write\n");
+	fprintf(stderr, "Path: %s\n", path);
+	int fd;
+	int res;
+
+	char fpath[PATH_MAX];
+	get_file_path(fpath, fname);
+	fprintf(stderr, "FPath: %s\n", fpath);
+
+	fd = open(fpath, O_CREAT | O_WRONLY, 0644);
+
+	if (fd == -1)
+		return -errno;
+
+	res = pwrite(fd, buf, size, offset);
```

Comment:
Any reason why pwrite was used instead of fwrite?

Reply:
So fwrite is a wrapper over pwrite but does not allow random offset seeking. Since we are implementing the base system call, pwrite is what allows you to seek a random offset and write. That is essentially the core system call.